src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -0,0 +1,421 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace System.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// TCP connection handle
+    /// </summary>
+    internal class SNITCPHandle : SNIHandle
+    {
+        private readonly string _targetServer;
+        private object _callbackObject;
+        private Stream _stream;
+        private Socket _socket;
+        private NetworkStream _tcpStream;
+        private SslStream _sslStream;
+        private SslOverTdsStream _sslOverTdsStream;
+        private TcpClient _tcpClient;
+        private SNIAsyncCallback _receiveCallback;
+        private SNIAsyncCallback _sendCallback;
+        private TaskScheduler _writeScheduler;
+        private TaskFactory _writeTaskFactory;
+
+        private bool _validateCert = true;
+        private int _bufferSize = 4096;
+        private uint _status = TdsEnums.SNI_UNINITIALIZED;
+
+        public Guid _connectionId = Guid.NewGuid();
+
+        /// <summary>
+        /// Dispose object
+        /// </summary>
+        public override void Dispose()
+        {
+            lock (this)
+            {
+                if (_sslOverTdsStream != null)
+                {
+                    _sslOverTdsStream.Dispose();
+                    _sslOverTdsStream = null;
+                }
+
+                if (_sslStream != null)
+                {
+                    _sslStream.Dispose();
+                    _sslStream = null;
+                }
+
+                if (_tcpClient != null)
+                {
+                    _tcpClient.Dispose();
+                    _tcpClient = null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Connection ID
+        /// </summary>
+        public override Guid ConnectionId
+        {
+            get
+            {
+                return _connectionId;
+            }
+        }
+
+        /// <summary>
+        /// Connection status
+        /// </summary>
+        public override UInt32 Status
+        {
+            get
+            {
+                return _status;
+            }
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="serverName">Server name</param>
+        /// <param name="port">TCP port number</param>
+        /// <param name="timerExpire">Connection timer expiration</param>
+        /// <param name="callbackObject">Callback object</param>
+        public SNITCPHandle(string serverName, int port, long timerExpire, object callbackObject)
+        {
+            _writeScheduler = new LimitedConcurrencyLevelTaskScheduler(1);
```

Comment:
Per my previous comment, this could be:
```C#
_writeScheduler = new ConcurrentExclusiveSchedulerPair(1).ConcurrentScheduler;
```
or because you're actually just trying to serialize execution, making everything exclusive, you could do:
```C#
_writeScheduler = new ConcurrentExclusiveSchedulerPair().ExclusiveScheduler;
```